subsys/bluetooth/mesh/net.c

```diff
@@ -108,36 +110,46 @@ K_MEM_SLAB_DEFINE(loopback_buf_pool,
 		  sizeof(struct loopback_buf),
 		  CONFIG_BT_MESH_LOOPBACK_BUFS, __alignof__(struct loopback_buf));
 
-static uint32_t dup_cache[CONFIG_BT_MESH_MSG_CACHE_SIZE];
-static int   dup_cache_next;
+static uint32_t dup_cache[CONFIG_BT_MESH_MSG_CACHE_SIZE]
+			 [CONFIG_BT_MESH_MSG_CACHE_SIZE];
+static struct {
+	uint8_t idx;
+	uint8_t next[CONFIG_BT_MESH_MSG_CACHE_SIZE];
+} dup;
 
 static bool check_dup(struct net_buf_simple *data)
 {
 	const uint8_t *tail = net_buf_simple_tail(data);
 	uint32_t val;
+	uint8_t idx;
 	int i;
 
 	val = sys_get_be32(tail - 4) ^ sys_get_be32(tail - 8);
 
-	for (i = 0; i < ARRAY_SIZE(dup_cache); i++) {
-		if (dup_cache[i] == val) {
+	idx = val % ARRAY_SIZE(dup_cache[0]);
```

Comment:
will test and update

Reply:
I pondered it carefully. It has nothing to do with testing. Your concern is right. This method is more likely to invalidate large-scale caches at the same time, and the additional decryption operations that it brings are compared to the previous iterative query. Not an order of magnitude.

Reply:
From this, a reverse-order iteration method is proposed, which I think may be better.

Reply:
I’m do some test, with one node as publish send with generic onoff get mesh message(dst=0xffff, ttl = 5, net_xmit=(8,20)), other ten nodes actor as generic onoff server model, once receive get mesh message, will send generic onoff status(ttl = 5, dst=get message src)


**BEFORE:**
![UdaW76NLlT](https://user-images.githubusercontent.com/26378606/195247700-dc39f2f3-3a76-4c3f-bd58-bbe8f4807e6b.jpg)

**AFTER:**

Reply:
Although this is a specific scenario, I think it is enough.

Reply:
You also need to consider a segmented message (preferably the biggest one) because 1) they will are not decrypted one by one, but once the whole message is received, and 2) the already received segment triggers the reassembly logic at the transport layer.

Reply:
But generally yes, the reverse order seems more reasonable.

Reply:
I don't quite understand why the influence of the transport layer should be considered. This is something of the network layer. It does not depend on the transport layer, because each message is encrypted with a different network nonce. I think the current test can explain the situation.